src/Pate/Verification/PairGraph.hs

```diff
@@ -0,0 +1,878 @@
+  ( Gas(..)
+  , PairGraph(..)
+  , initializePairGraph
+  , chooseWorkItem
+  , updateDomain
+  , pairGraphComputeFixpoint
+  , runVerificationLoop
+  ) where
+
+import qualified Control.Concurrent.MVar as MVar
+import           Control.Lens ( view, (^.), _1 )
+import           Control.Monad (foldM, when, forM, forM_, unless)
+import           Control.Monad.IO.Class
+import           Control.Monad.Reader (asks)
+import           Control.Monad.Writer (tell, execWriterT)
+import           Control.Monad.Except (runExceptT)
+import           Numeric (showHex)
+import           Prettyprinter
+
+import qualified Data.BitVector.Sized as BV
+import           Data.Maybe (fromMaybe)
+import           Data.Map (Map)
+import qualified Data.Map as Map
+import           Data.Proxy
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Word (Word32)
+
+import           Data.Parameterized.Classes
+import           Data.Parameterized.Some
+import qualified Data.Parameterized.TraversableF as TF
+--import qualified Data.Parameterized.Context as Ctx
+
+import qualified What4.Expr as W4
+import qualified What4.Interface as W4
+import qualified What4.Protocol.Online as W4
+import qualified What4.Protocol.SMTWriter as W4
+import           What4.SatResult (SatResult(..))
+
+import qualified Lang.Crucible.Backend as LCB
+import qualified Lang.Crucible.Backend.Online as LCBO
+import qualified Lang.Crucible.LLVM.MemModel as CLM
+import qualified Lang.Crucible.Simulator.RegValue as LCS
+import           Lang.Crucible.Simulator.SymSequence
+
+import qualified Data.Macaw.CFG as MM
+import qualified Data.Macaw.Symbolic as MS
+
+import qualified Pate.Abort as PAb
+import qualified Pate.Arch as PA
+import qualified Pate.Binary as PBi
+import qualified Pate.Block as PB
+import qualified Pate.Config as PCfg
+import qualified Pate.Discovery as PD
+--import qualified Pate.Equivalence.MemoryDomain as PEM
+import qualified Pate.Equivalence as PE
+import qualified Pate.Equivalence.Error as PEE
+import qualified Pate.Equivalence.MemoryDomain as PEM
+import qualified Pate.Equivalence.RegisterDomain as PER
+import qualified Pate.MemCell as PMc
+import qualified Pate.Monad.Context as PMC
+import           Pate.Equivalence as PEq
+import qualified Pate.Equivalence.Statistics as PESt
+import qualified Pate.Equivalence.EquivalenceDomain as PEE
+import           Pate.Monad
+import qualified Pate.Memory.MemTrace as MT
+--import qualified Pate.Monad.Context as PMC
+import qualified Pate.Proof.Instances as PPI
+import qualified Pate.Monad.Environment as PME
+import           Pate.Panic
+import qualified Pate.PatchPair as PPa
+import qualified Pate.SimState as PS
+import qualified Pate.Solver as PS
+import qualified Pate.SimulatorRegisters as PSR
+
+import qualified Pate.Verification.Domain as PVD
+import qualified Pate.Verification.Validity as PVV
+import qualified Pate.Verification.SymbolicExecution as PVSy
+
+
+newtype Gas = Gas Word32
```

Comment:
What decrements the gas? Can we document an anticipated range of reasonable values?

Reply:
Right now, gas is used when you do a widening step along a particular control-flow path.  So, it should only be consumed as you traverse loops or recursive call cycles.